api/http_client.go

```diff
@@ -115,6 +122,192 @@ func ExtractHeader(name string, dest *string) func(http.RoundTripper) http.Round
 	}
 }
 
+// GitHub servers return non-printable characters as their unicode code point values.
+// The values of \u0000 to \u001F represent C0 ASCII control characters and
+// the values of \u0080 to \u009F represent C1 ASCII control characters. These control
+// characters will be interpreted by the terminal, this behaviour can be used maliciously
+// as an attack vector, especially the control character \u001B. This function escapes
+// all non-printable characters between \u0000 and \u00FF so that the terminal will
+// not interpret them.
+func SanitizeASCIIControlCharacters(rt http.RoundTripper) http.RoundTripper {
+	return &funcTripper{roundTrip: func(req *http.Request) (*http.Response, error) {
+		res, err := rt.RoundTrip(req)
+		if err != nil ||
+			!jsonTypeRE.MatchString(res.Header.Get("Content-Type")) ||
+			res.ContentLength == 0 {
+			return res, err
+		}
+		var sanitized bytes.Buffer
+		err = replaceControlCharacters(res.Body, &sanitized)
+		if err != nil {
+			err = fmt.Errorf("ascii control characters sanitization error: %w", err)
+		}
+		res.Body.Close()
+		res.Body = io.NopCloser(&sanitized)
```

Comment:
One consequence of this middleware is that by default, all Bodies of every response will first be fully read into a Buffer before the middleware yields control to other middleware or to other response processing code. This prevents any kind of streaming processing of response JSON.

Would it be feasible to replace `resp.Body` with an `io.ReadCloser` that does escaping on the fly as it forwards reads to the original response body and returns the result? That way, every response from the server will be immediately available to the code that makes the request without this middleware potentially delaying that.

Reply:
I like that idea. I will investigate implementing it in that manner.